### ISSUE TITLE
a problem with Firebase SDK 6.16 

### ISSUE THREAD

**Issue by PhixDev** [firebase/quickstart-unity#829]
### [REQUIRED] Please fill in the following fields:

  * Unity editor version: 2019.3.9f1
  * Firebase Unity SDK version: 6.16
  * Source you installed the SDK: Unity Package Manager
  * Problematic Firebase Component: Database
  * Other Firebase Components in use: Firebase core, Firebase App, Firebase Auth, Firebase database
  * Additional SDKs you are using: unity ads
  * Platform you are using the Unity editor on: Windows
  * Platform you are targeting: for now android
  * Scripting Runtime: Mono

### [REQUIRED] Please describe the issue here:

My console gives me these errors... Please help me

![95366150-59a46600-08d3-11eb-88cb-69543622cce3](https://user-images.githubusercontent.com/72518593/96478088-8c395180-1237-11eb-8eaa-62d23637513a.png)

![95366224-717bea00-08d3-11eb-994e-efa7deb8ab85](https://user-images.githubusercontent.com/72518593/96478094-8fccd880-1237-11eb-87d1-96287c1a1e79.png)

I'm trying since several weeks to get help from the firebase support... but they don't help me :'

**Comment by ed-yaku** [comment]
Hi,

I have recently upgraded to Firebase Unity SDK 6.16.0 (and later to 6.16.1 once that was released) and I'm also seeing this error. Is there confirmation that this issue is being seen internally and what is the advised course of action?

I took the decision to upgrade after following the resolution from this issue https://github.com/firebase/quickstart-unity/issues/788 but I'm not confident sticking with this upgrade even if this issue is isolated to the Unity Editor.

Thanks

**Comment by PhixDev** [comment]
> Hi! I've attempted to reproduce this with the Realtime Database quickstart using Firebase Realtime Database 6.16.1 installed via the package manager. Unfortunately (?) everything worked as expected on Android.
> 
> Is there chance that you can attempt to alter the quickstart to reproduce this problem? Thanks!

what do you mean?

I've tried to build the game and play it on my smartphone, it works without problems, but when i try to play the game from the Unity Editor it gives me these errors when the script of the realtime database go in function.

**Comment by ed-yaku** [comment]
As @PhixDev says on Android and iOS the functionality is fine. This error is isolated to the Unity Editor.

**Comment by DellaBitta** [comment]
I wasn't able to reproduce this in Windows or the Mac Editor using the test app.  A couple of questions:

1. Are there any special characters in your database name? (I cannot see it since it's been stricken out in your logs, and I just want to make sure).
2. Have you don this step to programmatically set the Editor Database URL as specified in the section "[Configuring the SDK for the Unity Editor](https://firebase.google.com/docs/database/unity/start#configuring_the_sdk_for_the_unity_editor)"?

Thanks!

**Comment by ed-yaku** [comment]
Hi @DellaBitta thanks for testing that out.

1. The Database URL used while testing matches the following regex `https:\/\/([a-zA-Z0-9\-\.]+)\.firebaseio\.com\/`
2. For the Editor setup I've tried `FirebaseApp.DefaultInstance.SetEditorDatabaseUrl("https://MY-FIREBASE-APP.firebaseio.com/");` and `app.Options.DatabaseUrl = new Uri("https://MY-FIREBASE-APP.firebaseio.com/");` where `app` is the `FirebaseApp`. NB: the former is now obsolete ("Please use FirebaseApp.Options.DatabaseUrl instead"). All the other `SetEditor` methods are also obsolete and shouldn't be used.

I have now had to downgrade to 6.15.2, where the Editor Database functionality still works, as I can't spend anymore time on the task.

**Comment by chkuang-g** [comment]
@PhixDev 

Just to confirm, when this happens, the urls in the log are DIFFERENT, correct?

The error message is from [here](https://github.com/firebase/firebase-cpp-sdk/blob/f0a9d5b65e3a014aee347ea9bb3b664cdce2ae55/database/src/desktop/database_desktop.cc#L140) when the code try to compare the passed in url from `GetReferenceFromUrl()` API and the one cached in the database instance. The database instance will cache the url stored in App upon creation.

Unfortunately, `FirebaseApp.DefaultInstance.SetEditorDatabaseUrl()` is deprecated and will be removed in the next release. And I do not think you can modify `app.Options.DatabaseUrl` after app is created.

Instead, if you need a different database url for your test in the editor, try to utilize `FirebaseApp.Create( AppOptions options)`
https://firebase.google.com/docs/reference/unity/class/firebase/firebase-app#class_firebase_1_1_firebase_app_1aaeb3418d9c118fcb63a409ee05a5e4aa

For instance, if the database is in the same Firebase project, you can simply do the follolwing:
```
   AppOptions options = new AppOptions();
   options.DatabaseUrl = "https://MY-FIREBASE-APP.firebaseio.com/";
   FirebaseApp app = FirebaseApp.Create(options)
```

If it is in a different Firebase project, you will need to add API keys and AppId, etc.
https://firebase.google.com/docs/reference/unity/class/firebase/app-options#class_firebase_1_1_app_options

**Comment by PhixDev** [comment]
> @PhixDev 
> 
> Just to confirm, when this happens, the urls in the log are DIFFERENT, correct?
> 
> The error message is from [here](https://github.com/firebase/firebase-cpp-sdk/blob/f0a9d5b65e3a014aee347ea9bb3b664cdce2ae55/database/src/desktop/database_desktop.cc#L140) when the code try to compare the passed in url from `GetReferenceFromUrl()` API and the one cached in the database instance. The database instance will cache the url stored in App upon creation.
> 
> Unfortunately, `FirebaseApp.DefaultInstance.SetEditorDatabaseUrl()` is deprecated and will be removed in the next release. And I do not think you can modify `app.Options.DatabaseUrl` after app is created. Even you modify it, the url may have been cached in the database instance already.
> 
> Instead, if you need a different database url for your test in the editor, try to utilize `FirebaseApp.Create( AppOptions options)`
> https://firebase.google.com/docs/reference/unity/class/firebase/firebase-app#class_firebase_1_1_firebase_app_1aaeb3418d9c118fcb63a409ee05a5e4aa
> 
> For instance, if the database is in the same Firebase project, you can simply do the follolwing:
> ```
>    AppOptions options = new AppOptions();
>    options.DatabaseUrl = "https://MY-FIREBASE-APP.firebaseio.com/";
>    FirebaseApp app = FirebaseApp.Create(options)
> ```
> 
> If it is in a different Firebase project, you will need to add API keys and AppId, etc.
> https://firebase.google.com/docs/reference/unity/class/firebase/app-options#class_firebase_1_1_app_options
> 
> 

Hi, I need to try better but I think the URLs into the logs are same.

**Comment by ed-yaku** [comment]
To confirm from my case the URLs are exactly the same in the error log.

**Comment by DellaBitta** [comment]
The error logger in that bit of code unfortunately doesn't log the parsed string from the hostInfo.ToString() but the string passed into the method itself.  Therefore the output may be the same in the log but the logical comparison could fail.  Maybe there's an issue there where the ToString() method isn't properly handling some cases?

I could add code to output the ToString() result in the log as well, and I will add ToString() comparison tests to the suite, but it would be better if we could figure this out beforehand if possible.

Could you please describe the string?  The length, any symbolic characters, etc?  From the original post in this ticket it seems to be a https url, is that true for everyone else, too?  We do have [some tests for the URL parser](https://github.com/firebase/firebase-cpp-sdk/blob/1d13237b37dcecdc3a48ccd882337e03d4ac5bf0/database/tests/desktop/util_desktop_test.cc#L1234).  Does it match any of those patterns?

Thanks.

**Comment by ed-yaku** [comment]
Hi @DellaBitta sorry for the delay getting back to you as I was on leave.

My URL is of the form `https://test-123.firebaseio.com/`

**Comment by DellaBitta** [comment]
I've created an internal ticket b/172233676 to track the work to investigate this issue and I'm going to mark it as a bug for now.

**Comment by DellaBitta** [comment]
Hi @ed-yaku and @PhixDev,

I'm still unable to reproduce this problem either in the editor (hitting play) or by creating a targeted standalone windows build and running the program there.  Could you please paste the code that you're using to define the database URL in the configuration, or is it simply being read directly from the google-services.json file?

Also, out of curiosity, what happens if you take away the trailing "/" character at the end of the URL?

**Comment by ed-yaku** [comment]
Hi @DellaBitta as an update I've tested using the same database URL in a stripped back project and I'm unable to reproduce the issue. 

In the project where the issue is happening it occurs if I've defined the URL or just used the data from the google-services-desktop.json file in the StreamingAssets folder.

I'll throw some more time next week to see if I can find the reproduction case in the stripped back project.
